vllm/core/block/prefix_caching_block.py

```diff
@@ -398,6 +473,24 @@ def is_full(self) -> bool:
     def num_empty_slots(self) -> int:
         return self._block.num_empty_slots
 
+    @property
+    def num_tokens_total(self) -> int:
+        """return the total tokens so far.
+
+        Here we iterate the block chain till to the first block, while
+        cache the result in local to prevent repeated computations.
+        """
+        if self._cached_num_tokens_total is not None:
+            return self._cached_num_tokens_total
+
+        _block = self
+        self._cached_num_tokens_total = 0
+        while _block is not None:
+            self._cached_num_tokens_total += len(_block.token_ids)
+            _block = _block.prev_block
```

Comment:
This is still O(N^2) for the entire sequence, e.g. each block will perform O(N) scan. We should use the design I linked to you [`content_hash`](https://github.com/vllm-project/vllm/blob/4bb53e2dde809ea5727b8cac95a080893733a1ef/vllm/core/block/prefix_caching_block.py#L414) which is O(N).

I will not block the PR (we can optimize later) but you should add a TODO!

Reply:
I see, I would add this TODO note